cirq/testing/consistent_specified_has_unitary.py

```diff
@@ -0,0 +1,26 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any
+
+from cirq import protocols
+
+
+def assert_specifies_has_unitary_if_unitary(val: Any) -> None:
+    """Checks that unitary values can be cheaply identifies as unitary."""
+    assert not protocols.has_unitary(val) or hasattr(val, '_has_unitary_'), (
```

Comment:
It would be worth verifying that `_has_unitary_` didn't just return `NotImplemented`.

Reply:
Is there a test that checks `cirq.unitary(val, default=None) is None == cirq.has_unitary(val)`?

If not, this test could be added here.

Reply:
Eh, if `_has_unitary_` returned `NotImplemented` then at least there's code intentionally triggering a particular behavior.

Yes, there is a test checking consistency between the unitary and has_unitary methods.